src/nias_cpp/interfaces/inner_products.h

```diff
@@ -0,0 +1,125 @@
+#ifndef NIAS_CPP_INTERFACES_INNER_PRODUCTS_H
+#define NIAS_CPP_INTERFACES_INNER_PRODUCTS_H
+
+#include <algorithm>
+#include <memory>
+#include <optional>
+
+#include <nias_cpp/concepts.h>
+#include <nias_cpp/indices.h>
+#include <nias_cpp/interfaces/vectorarray.h>
+#include <nias_cpp/type_traits.h>
+#include <pybind11/numpy.h>
+
+namespace nias
+{
+
+
+template <floating_point_or_complex F>
+class SesquilinearFormInterface
+{
+   public:
+    using ScalarType = F;
+
+    SesquilinearFormInterface() = default;
+    SesquilinearFormInterface(const SesquilinearFormInterface&) = default;
+    SesquilinearFormInterface(SesquilinearFormInterface&&) = default;
+    SesquilinearFormInterface& operator=(const SesquilinearFormInterface&) = default;
+    SesquilinearFormInterface& operator=(SesquilinearFormInterface&&) = default;
+    virtual ~SesquilinearFormInterface() = default;
+
+    /**
+     * \brief Apply the sesquilinear form
+     *
+     * If \c pairwise is \c true, the form is applied to each pair of vectors in the arrays and
+     * the result has the same length as the input arrays (which have to have the same size in this case).
+     * If \c pairwise is \c false, the form is applied to each vector in the first array with each vector
+     * of the second array, and the result has size <tt>left.size() * right.size()</tt>
+     * (where the <tt>i * right.size() + j</tt>-th element contains the result of applying the form to
+     * the <tt>i</tt>-th vector of the left array with the <tt>j</tt>-th vector of the right array).
+     */
+    [[nodiscard]] virtual std::vector<ScalarType> apply(
```

Comment:
I'd still go for a `std::vector<std::vector<ScalarType>>` here, to distinguish between `apply_pairwise`.

Reply:
Done.